core/tracing/src/main/scala/org/typelevel/otel4s/trace/Span.scala

```diff
@@ -0,0 +1,233 @@
+  *
+  * Properly ended span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createManual.use { span =>
+  *     span.setStatus(Status.Ok, "all good") >> span.end
+  *   }
+  * }}}
+  *
+  * ==[[Span.Auto]]==
+  * Unlike [[Span.Manual]] the auto span has a fully managed lifecycle. That
+  * means the span is started upon resource allocation and ended upon
+  * finalization.
+  *
+  * Automatically ended span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createAuto.use { span =>
+  *     span.setStatus(Status.Ok, "all good")
+  *   }
+  * }}}
+  *
+  * ==[[Span.Res]]==
+  * The behaviour and lifecycle management are identical to [[Span.Auto]]. The
+  * allocation and release stages of a supplied resource are traced by separate
+  * spans. Carries a value of a wrapped resource.
+  *
+  * The structure of the inner spans:
+  * {{{
+  * > span-name
+  *   > acquire
+  *   > use
+  *   > release
+  * }}}
+  *
+  * Resource span:
+  * {{{
+  * val tracer: Tracer[F] = ???
+  * val resource: Resource[F, String] = Resource.eval(Sync[F].delay("string"))
+  * val ok: F[Unit] =
+  *   tracer.spanBuilder("manual-span").createRes(resource).use { case span @ Span.Res(value) =>
+  *     span.setStatus(Status.Ok, s"all good. resource value: $${value}")
+  *   }
+  * }}}
+  */
+trait Span[F[_]] {
+
+  /** Returns the [[SpanContext]] associated with this span.
+    */
+  def context: SpanContext
+
+  /** Sets an attribute to the span. If the span previously contained a mapping
+    * for the key, the old value is replaced by the specified value.
+    *
+    * @param attribute
+    *   the attribute to add to the span
+    */
+  def setAttribute[A](attribute: Attribute[A]): F[Unit]
+
+  /** Sets attributes to the span. If the span previously contained a mapping
+    * for any of the keys, the old values are replaced by the specified values.
+    *
+    * @param attributes
+    *   the set of attributes to add to the span
+    */
+  def setAttributes(attributes: Attribute[_]*): F[Unit]
+
+  /** Adds an event to the span with the given attributes. The timestamp of the
+    * event will be the current time.
+    *
+    * @param name
+    *   the name of the event
+    *
+    * @param attributes
+    *   the set of attributes to associate with the event
+    */
+  def addEvent(name: String, attributes: Attribute[_]*): F[Unit]
+
+  /** Adds an event to the span with the given attributes and timestamp.
+    *
+    * '''Note''': the timestamp should be based on `Clock[F].realTime`. Using
+    * `Clock[F].monotonic` may lead to an incorrect data.
+    *
+    * @param name
+    *   the name of the event
+    *
+    * @param timestamp
+    *   the explicit event timestamp since epoch
+    *
+    * @param attributes
+    *   the set of attributes to associate with the event
+    */
+  def addEvent(
+      name: String,
+      timestamp: FiniteDuration,
```

Comment:
Might a `java.time.Instant` be more appropriate here?

Reply:
`FiniteDuration` fits well with `Clock[F].realTime`.

Also, it should be possible to publish a Scala Native version too. The implementation remains questionable, but at least it's possible.

Reply:
> The implementation remains questionable, but at least it's possible.

Yes, we will probably want a pure Scala+CE backend/implementation at some point anyway. Seems silly to work so hard on I/O integrated runtimes when users just drag in Java libs with threadpools and such 😂

Reply:
That's a misuse of `FiniteDuration`, but consistency with `Clock` is sensible, and it'll be somebody else's problem in 2262 when it overflows.  I sigh wearily and concede.